org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/handlers/SaveActionHandler.java

```diff
@@ -155,4 +167,26 @@ private List<TextEdit> handleSaveActionOrganizeImports(String documentUri, IProg
 		return edit;
 	}
 
+	private void handleSaveActionRenameFile(String documentUri, IProgressMonitor monitor) {
+		ICompilationUnit cu = JDTUtils.resolveCompilationUnit(documentUri);
+		CompilationUnit astRoot = CoreASTProvider.getInstance().getAST(cu, CoreASTProvider.WAIT_YES, null);
+		IProblem[] problems = astRoot.getProblems();
+		Optional<IProblem> desiredProblem = Arrays.stream(problems).filter(p -> p.getID() == IProblem.PublicClassMustMatchFileName).findFirst();
+		if (desiredProblem.isPresent()) {
+			IProblem renameProblem = desiredProblem.get();
+			String newName = renameProblem.getArguments()[1];
```

Comment:
I was trying to see if there's a nicer way to do this, because this basically relies on the fact that the problem contains the new name in its error message of `The public type ${typeName} must be defined in its own file`

You could use :
`((org.eclipse.jdt.core.dom.TypeDeclaration)astRoot.types().get(0)).getName().getIdentifier()`.

However I'm not sure if it's safe to assume that the primary type will always be the 0th element. Also this complicates things for records (not a type declaration) so I think this is fine for now.